serde_list/serde_list_macros/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+pub fn derive_externally_tagged(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+    let enum_name = ast.ident;
+
+    let Data::Enum(data) = ast.data else {
+        panic!();
+    };
+
+    let mut valid = false;
+    for attr in &ast.attrs {
+        if attr.path().is_ident("repr") {
+            if attr.parse_args::<Ident>().unwrap() == "u8" {
+                valid = true;
+            }
+        }
+    }
+    if !valid {
+        panic!("ExternallyTagged requires #[repr(u8)]")
+    }
+
+    let mut field_num_impls = Vec::new();
+    let mut field_serialize_impls = Vec::new();
+
+    for variant in data.variants {
+        let variant_name = variant.ident;
+
+        let fields = match variant.fields {
+            Fields::Named(fields_named) => fields_named.named,
+            Fields::Unnamed(_fields_unnamed) => panic!(),
+            Fields::Unit => Punctuated::new(),
+        };
+
+        let mut names = Vec::new();
+        let mut serialized = Vec::new();
+
+        for field in &fields {
+            let Some(field_name) = &field.ident else {
+                panic!();
+            };
+
+            names.push(field_name);
+            serialized.push(quote! { state.serialize_element(#field_name)?; });
+        }
+
+        let num = names.len();
+
+        field_num_impls.push(quote! {
+            Self::#variant_name { .. } => { #num },
+        });
+
+        field_serialize_impls.push(quote! {
+            Self::#variant_name { #(#names),* } => { #(#serialized)* },
+        });
+    }
+
+    quote! {
+        impl ExternallyTagged for #enum_name {
+            // https://doc.rust-lang.org/reference/items/enumerations.html?search=#pointer-casting
+            fn discriminant(&self) -> u8 {
+                // This is safe if the enum has repr(u8)
+                let pointer = self as *const Self as *const u8;
+                unsafe { *pointer }
+            }
+
+            fn num_fields(&self) -> usize {
+                match self {
+                    #(#field_num_impls)*
+                }
+            }
+
+            fn serialize_fields<S: serde::ser::SerializeSeq>(&self, state: &mut S) -> Result<(), S::Error> {
+                Ok(match self {
+                    #(#field_serialize_impls)*
+                })
+            }
+        }
+    }
+    .into()
+}
+
+#[proc_macro_derive(Serialize_custom_u8)]
+pub fn derive_serialize_custom_u8(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+    let enum_name = ast.ident;
+
+    let Data::Enum(data) = ast.data else {
+        panic!();
+    };
+
+    let mut valid = false;
+    for attr in &ast.attrs {
+        if attr.path().is_ident("repr") {
+            if attr.parse_args::<Ident>().unwrap() == "u8" {
+                valid = true;
+            }
+        }
+    }
+    if !valid {
+        panic!("Serialize_custom_u8 requires #[repr(u8)]")
+    }
```

Comment:
Can you put the logic above into a separate function? Then you can re-use it across the two macros.